src/main/java/pl/cleankod/exchange/provider/FindAccountService.java

```diff
@@ -0,0 +1,48 @@
+package pl.cleankod.exchange.provider;
+
+import pl.cleankod.exchange.core.domain.Account;
+import pl.cleankod.exchange.core.usecase.FindAccountAndConvertCurrencyUseCase;
+import pl.cleankod.exchange.core.usecase.FindAccountUseCase;
+import pl.cleankod.util.AccountFailedReason;
+import pl.cleankod.util.Preconditions;
+import pl.cleankod.util.Result;
+
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Currency;
+import java.util.Optional;
+
+public class FindAccountService {
+    private final FindAccountAndConvertCurrencyUseCase findAccountAndConvertCurrencyUseCase;
+    private final FindAccountUseCase findAccountUseCase;
+
+    public FindAccountService(
+            FindAccountAndConvertCurrencyUseCase findAccountAndConvertCurrencyUseCase,
+            FindAccountUseCase findAccountUseCase
+    ) {
+        this.findAccountAndConvertCurrencyUseCase = findAccountAndConvertCurrencyUseCase;
+        this.findAccountUseCase = findAccountUseCase;
+    }
+
+    public Result<Account, AccountFailedReason> findAccountById(String id, String currency) {
+        Preconditions.requireNonNull(id);
+
+        Account.Id accountId = Account.Id.of(URLDecoder.decode(id, StandardCharsets.UTF_8));
+        return Optional.ofNullable(currency)
+                .map(s -> findAccountAndConvertCurrencyUseCase.execute(accountId, Currency.getInstance(s)))
+                .orElseGet(() -> findAccountUseCase.execute(accountId))
```

Comment:
Having `Optional<Optional<Account>>` is odd. I'd suggest making a `flatMap()` and `or()`:
```java
return Optional.ofNullable(currency)
        .flatMap(s -> findAccountAndConvertCurrencyUseCase.execute(accountNumber, Currency.getInstance(s)))
        .or(() -> findAccountUseCase.execute(accountNumber))
        .map(Result::<Account, AccountFailedReason>successful)
        .orElseGet(() -> Result.fail(AccountFailedReason.NOT_FOUND));
```